### PR TITLE
 Make JIT tracing a thread-local property

### PR DESCRIPTION
**Pull request by apaszke**
As in the title. Lets us simplify a lot of code.

Depends on #9363, so please review only the last commit.

@zdevito

### CODE REVIEW

**Comment by anderspapitto** [comment]
thread locals have been causing trouble for me over at https://github.com/pytorch/pytorch/pull/8792 because once python-less libtorch.so is split out from python-full _C.so, Windows doesn't support exposing thread_local variables across the shared-object boundary. Will this PR introduce another occurrence of this issue?

**Comment by SsnL** [comment]
@ezyang told me that this will solve #9069 :D

**Comment by ezyang** [comment]
@anderspapitto The thread local is not exported (it is not even in a header) so there should be no problem.

File: tools/autograd/templates/VariableType.cpp [hunk 1]
```diff
@@ -38,10 +38,14 @@ using namespace torch::autograd::generated;
 namespace torch { namespace autograd {
 // Helper methods for working with Attributes (torch/csrc/jit/attributes.h)
 
+at::Tensor maybeUnwrapVar(const at::Tensor& t) {
```
Comment (ezyang) [inline tools/autograd/templates/VariableType.cpp]: This seems a bit awkward. What was the proximal cause for this?

File: torch/csrc/Size.cpp [hunk 2]
```diff
@@ -41,7 +41,7 @@ PyObject * THPSize_NewFromSizes(int dim, const int64_t *sizes)
 static bool isTracedVar(PyObject *item) {
   if (!THPVariable_Check(item)) return false;
   auto & var = reinterpret_cast<THPVariable*>(item)->cdata;
-  return torch::jit::tracer::isTracing(var);
+  return var.dim() == 0 && torch::jit::tracer::getValueTrace(var);
```
Comment (ezyang) [inline torch/csrc/Size.cpp]: dim() == 0?

File: torch/csrc/jit/tracer.cpp [hunk 3]
```diff
@@ -38,35 +48,39 @@ void postRecordTrace(const PreTraceInfo& info,
...
 
   return size_var;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// Argument stash
+////////////////////////////////////////////////////////////////////////////////
+thread_local ArgumentStash ArgumentStash::stash;
+
+void ArgumentStash::stashIntListElem(const std::string& arg_name, size_t size, size_t idx, const Variable& var) {
```
Comment (ezyang) [inline torch/csrc/jit/tracer.cpp]: Do we still need the stasher now that we have a thread local tracer? (Not sure, but wondering.)

**Review by ezyang — APPROVED** [review]
seems legit

File: tools/autograd/templates/VariableType.cpp (hunk 1 above)
Comment (apaszke) [reply tools/autograd/templates/VariableType.cpp]: This fixes #8755. The problem was that once we call `.toTensor()` on `at::Scalar`s below then we usually get a non-variable tensor, but we recently made a switch to use variables everywhere.

File: torch/csrc/Size.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/Size.cpp]: Yeah, we want to allow putting tensors into `torch.Size`s, but they need to be scalars

File: torch/csrc/jit/tracer.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/jit/tracer.cpp]: I could move this function to `TracingState` if you think that would be nicer, but unfortunately we can't get rid of that. It's our way of retaining the information that was present in tensors that got casted into `IntList`s along their way to ATen.

**Review by zdevito — APPROVED** [review]
This looks good! Cleans up a lot of stuff.

**Comment by apaszke** [comment]
I would change the name of the function, but this would mean a whole roundtrip through the internal CI, which I'd like to avoid. We can fix it up later.

**Pull request closed without merging**